tests/test_operations.py

```diff
@@ -591,3 +591,29 @@ def test_formatted_output_unicode():
     with state.settings(line_template=line_template):
         result = operations._print_line(StringIO(), unicode_point)
         assert expected_stdout == result
+
+
+def test_rsync_upload_command():
+    expected = "rsync --rsh='ssh -i /example/path/id_rsa -p 23 -o StrictHostKeyChecking=no' /local/foo elife@34.234.82.201:/remote/bar"
+    with state.settings(
+        user="elife",
+        port=23,
+        host_string="34.234.82.201",
+        key_filename="/example/path/id_rsa",
+    ):
+        assert expected == " ".join(
```

Comment:
remove the `' '.join` stuff here.

the _rsync_upload function should return the final command to be used with `local` with nothing further to do